### PR TITLE
Added rigid to rope constraint

### PR DESCRIPTION
**Pull request by Scottyprogrammer**
We NEED to control rigidity. I saw that Shadow tried getting this in but not sure why it hasn't.

### CODE REVIEW

**Comment by Divran** [comment]
Hm. This will break all old duplicated E2s that use that function. Maybe it would be better to overload the function, so that we have two: one with and one without this new argument. What does everyone else think @AbigailBuccaneer @TomyLobo @Nebual

**Comment by CaptainPRICE** [comment]
Yep. Overload it instead.

**Comment by Scottyprogrammer** [comment]
Added a new one.

**Comment by Scottyprogrammer** [comment]
Tested and finalized. Working great.

**Comment by TomyLobo** [comment]
@Divran you ignored me when you added constraintcore in the first place, why bother asking now?

**Comment by Divran** [comment]
I didn't add it though, that was nebual

**Comment by Scottyprogrammer** [comment]
I refer the e:function() method as well, but its not that bad. It does make sense when you have function( ent1, vec1, ent2, vec2 ) though.

**Comment by Scottyprogrammer** [comment]
Lets get these added though. All the e2s I made before I took a break from gmod relied on Omi's constraint core, and the official version doesn't have all the functions needed.

**Comment by Divran** [comment]
giving @TomyLobo a chance to reply first but I don't see any reason not to merge

**Comment by Scottyprogrammer** [comment]
Sweet, I'm excited. Haven't played gmod in a while and to get back into it will be fun. Time to rewrite my chips.

**Comment by TomyLobo** [comment]
I think weld(entity,entity) is better because it's a symmetric relationship.

having two entity/vector pairs at the start of the parameter list for ropes is fine too.
But what is that reference id?

**Comment by Scottyprogrammer** [comment]
The reference ids allow you to modify the ropes' add length as seen in the SetLength funtion.

**Comment by TomyLobo** [comment]
What do you mean by "originally"? You mean you want to make a 1:1 mapping of Garry's functions?
I think making a good interface is a better idea :)

**Comment by Scottyprogrammer** [comment]
Originally as in the e2 rope functions had index in them when constraint core was written.

**Comment by Scottyprogrammer** [comment]
How would you suggest referencing the ropes and/or hydraulics instead of an index stored on the attached entity?

**Comment by TomyLobo** [comment]
with their own type, of course. That's much more flexible.

File: lua/entities/gmod_wire_expression2/core/custom/constraintcore.lua [hunk 1]
```diff
@@ -166,6 +166,22 @@ e2function void rope(index, entity ent1, vector v1, entity ent2, vector v2, addl
...
 
+--- Creates a rope between <ent1> and <ent2> at vector positions local to each ent, with <addlength> additional length, <width> width, and <mat> material.
+e2function void rope(index, entity ent1, vector v1, entity ent2, vector v2, addlength, width, string mat, rigid )
+	if not checkEnts(self, ent1, ent2) then return end
+	if not ent1.data then ent1.data = {} end
+	if not ent1.data.Ropes then ent1.data.Ropes = {} end
+	local vec1, vec2 = Vector(v1[1], v1[2], v1[3]), Vector(v2[1], v2[2], v2[3])
+	local length = (ent1:LocalToWorld(vec1) - ent2:LocalToWorld(vec2)):Length()
+	
+	if IsValid(ent1.data.Ropes[index]) then
```
Comment (TomyLobo) [inline lua/entities/gmod_wire_expression2/core/custom/constraintcore.lua]: Factoring some of this out into a helper function would be nice, in order to reduce copy-pasted errors.

**Comment by Divran** [comment]
>with their own type, of course. That's much more flexible.
>@Divran how do we handle this?

It's a bit too late for that now. Nebual's constraint core has been in wire for a long time so we would break everyone's E2s if we changed it to use custom data types. This PR only adds one more argument to a function that already exists, so I say just merge

**Comment by TomyLobo** [comment]
Well we can keep the old functions as deprecated

**Comment by Divran** [comment]
Meh, I don't see it as that big of a deal. I say we merge all of these few PRs and then if someone wants to remake it later they still can.

**Comment by TomyLobo** [comment]
but adding new features would provide an incentive to use the new system

**Comment by Divran** [comment]
Yes but adding new features takes longer than pressing a button, so there's no reason to not merge these in the meantime

**Comment by oldmud0** [comment]
tldr: MERGE IT!!!1

**Comment by TomyLobo** [comment]
@Divran But I just named a reason, how can you say there's no reason?

**Comment by Scottyprogrammer** [comment]
The additions I'm making are the incentive for me use get back into good and use constraint core. Currently constraint core is useless for my needs because it doesn't have rigidity/constant/damping/etc.

**Comment by Divran** [comment]
So @TomyLobo, we talked a bit on steam, shall I merge it now?

**Comment by Nebual** [comment]
Constraints are mostly between multiple entities, and switching to E:constraint(E2) implies one of them owns the constraint, which isn't overly accurate. 

Pull Requests shouldn't be held back for changing too few lines of code, this one purely adds functionality while following existing conventions, so I'd say merge it.

**Comment by TomyLobo** [comment]
go ahead
